api/envoy/config/core/v3/protocol.proto

```diff
@@ -255,11 +255,30 @@ message HttpProtocolOptions {
   // <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.drain_timeout>`.
   google.protobuf.Duration max_connection_duration = 3;
 
-  // The maximum number of headers. If unconfigured, the default
-  // maximum number of request headers allowed is 100. Requests that exceed this limit will receive
-  // a 431 response for HTTP/1.x and cause a stream reset for HTTP/2.
+  // The maximum number of headers (request headers if configured on HttpConnectionManager,
```

Comment:
It seems a little odd to have one knob that is used for request headers in one context but response headers in another context.  Maybe this should be split into two knobs?

Reply:
I was following the pre-existing pattern for `max_headers_count` for consistency. If we don't like this pattern, what's a better way to do this? I think the Envoy architecture means that the request limit needs to be somewhere on the HCM, and the response limit needs to be on the Cluster, and this message is used in both places currently.

Reply:
Just to be clear, I'm not changing the behavior of `max_headers_count` in this PR; I'm just improving the documentation on what it actually does, but it has always behaved this way.

Reply:
I wasn't proposing any architectural changes in terms of where Envoy enforces these limits; I was just proposing that we have separate knobs for request and response headers (e.g., `max_request_headers_count` and `max_response_headers_count`).

I realize that for Envoy, outbound limits will always make sense in the HCM and inbound limits will always make sense in the cluster, but that may not be true of other data planes that use xDS.  For example, if we ever need to support these limits in gRPC, the gRPC client would probably work the same way that Envoy does, but the gRPC server (which doesn't use CDS at all) would probably need to enforce both inbound and outbound limits in the HCM, and it would be impossible to configure the two limits independently if there's only a single knob.

Thanks for pointing out that `max_headers_count` already works this way; I wasn't aware of that.  Given that, let me suggest the following:

1. Add new `max_request_headers_count` and `max_response_headers_count` fields.  If either of those fields are set, they take precedence over the existing `max_headers_count` field for the specified limits.  Individual data planes can document which limits they enforce in which contexts (e.g., Envoy can ignore `max_response_headers_count` in the HCM).

2. For the new field, instead of adding `max_headers_kb`, instead add two new fields, `max_request_headers_kb` and `max_response_headers_kb`.  Since the combined field didn't already exist, we don't need a backward-compatible story for this one.

Thoughts...?

Reply:
Adding both fields (request and response variants) makes it more complicated in an envoy config because now there's always a field that isn't valid in the context it's in, which seems more error-prone. But OTOH, it does make it clearer what the settings means/does.

I think fixing `max_headers_count` is outside the scope of this PR, but I could add this as `max_response_headers_kb`; there's already an existing HCM setting for request headers kb. That would keep this PR from making the current situation any worse, at the cost of a small amount of inconsistency (but it'll be well documented).

Reply:
Okay, that seems like a reasonable compromise.  That makes it easier for us to later add a `max_request_headers_kb` field in the future if we ever need one.

Thanks!

Reply:
I made this change. PTAL.

Reply:
Thanks, looks great!